livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/assistant_llm.py

```diff
@@ -0,0 +1,246 @@
+                        )
+                    ]
+                )
+            )
+
+        async def on_tool_call_created(self, tool_call: ToolCall):
+            print(f"\nassistant > {tool_call.type}\n", flush=True)
+
+        async def on_tool_call_delta(self, delta: ToolCallDelta, snapshot: ToolCall):
+            if delta.type == "code_interpreter":
+                pass
+
+    def __init__(
+        self,
+        *,
+        cache_key: int,
+        client: AsyncClient,
+        sync_openai_task: asyncio.Task[AssistantLoadOptions],
+        chat_ctx: llm.ChatContext,
+        fnc_ctx: llm.FunctionContext | None,
+    ) -> None:
+        super().__init__(chat_ctx=chat_ctx, fnc_ctx=fnc_ctx)
+        self._cache_key = cache_key
+        self._client = client
+        self._thread_id: str | None = None
+        self._assistant_id: str | None = None
+
+        # current function call that we're waiting for full completion (args are streamed)
+        self._tool_call_id: str | None = None
+        self._fnc_name: str | None = None
+        self._fnc_raw_arguments: str | None = None
+        self._output_queue = asyncio.Queue[llm.ChatChunk | Exception | None]()
+        self._create_stream_task = asyncio.create_task(self._create_stream())
+        self._sync_openai_task = sync_openai_task
+
+        # Running stream is used to ensure that we only have one stream running at a time
+        self._running_stream: asyncio.Future[None] = asyncio.Future()
+        self._running_stream.set_result(None)
+
+    async def _create_stream(self) -> None:
+        await self._running_stream
+        self._running_stream = asyncio.Future[None]()
+        try:
+            load_options = await self._sync_openai_task
+            self._thread_id = load_options.thread_id
+            self._assistant_id = load_options.assistant_id
+
+            # At the chat_ctx level, create a map of thread_id to message_ids
+            # This is used to keep track of which messages have been added to the thread
+            # and which we may need to delete from OpenAI
+            if MESSAGES_ADDED_KEY not in self._chat_ctx._metadata:
+                self._chat_ctx._metadata[MESSAGES_ADDED_KEY] = dict[str, set[str]()]()
+
+            if self._thread_id not in self._chat_ctx._metadata[MESSAGES_ADDED_KEY]:
+                self._chat_ctx._metadata[MESSAGES_ADDED_KEY][self._thread_id] = set()
+
+            added_messages_set: set[str] = self._chat_ctx._metadata[MESSAGES_ADDED_KEY][
+                self._thread_id
+            ]
+            # Note: this will add latency unfortunately. Usually it's just one message so we loop it but
+            # it will create an extra round trip to OpenAI before being able to run inference.
+            for msg in self._chat_ctx.messages:
+                msg_id = msg._metadata.get(MESSAGE_ID_KEY)
+                if msg_id and msg_id not in added_messages_set:
+                    await self._client.beta.threads.messages.delete(
```

Comment:
can we run a task to sync all the changes that needed to be made vs await one by one?

Reply:
There isn't a way to order messages so it was one by one. BUT!

I just made an adjustment by using the `additional_messages` which creates the messages at stream time. This introduced a fair amount of additional complexity because now we don't know the open_ai message_id up front. But definitely worth it for latency.